src/System.IO.Pipelines/ReadableBufferReader.cs

```diff
@@ -80,7 +86,7 @@ public int Take()
 
         private void MoveNext()
         {
-            if (_enumerator.MoveNext())
+            if (_enumerator.MoveNext() && !_enumerator.Current.IsEmpty)
```

Comment:
I think you need to merge this but file and issue and assign it to me to fix it and add tests, I'd like to keep turnaround shorter.